torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -381,6 +381,48 @@ std::future<bool> launchAsyncGilCheck() {
   return resultFuture;
 }
 
+class CUDAEventCache {
+ public:
+  CUDAEventCache() : caches_(at::cuda::device_count()) {}
+  std::shared_ptr<at::cuda::CUDAEvent> create(int device, bool timing) {
+    auto& deviceCache = caches_[device];
+    auto deleter = [this, device, timing](at::cuda::CUDAEvent* event) {
+      auto& deviceCache = caches_[device];
+      std::lock_guard<std::mutex> lock(deviceCache.mutex);
+      deviceCache.events[timing ? 1 : 0].push_back(event);
+    };
+    at::cuda::CUDAEvent* event = nullptr;
+    {
+      std::lock_guard<std::mutex> lock(deviceCache.mutex);
+      auto& events = deviceCache.events[timing ? 1 : 0];
```

Comment:
Will deviceCache.events[timing ? 1 : 0] return only 1 'event' based on timing? if so, why plurals?

Reply:
deviceCache.events is two vector: a list of unused events without timing, and a list of unused events with timing. Since timing isn't strictly a global property (you can force timing on for a particular process group), we need to handle the situation where we fulfill both.

Reply:
Oh, I got it now, maybe it is just to me, but this was the confusing part to me: std::vector<at::cuda::CUDAEvent*> events[2], it is using a combination of vector and array semantics. Maybe a vector of vector is better.